### ISSUE TITLE
Are there any ways to make file transfers faster through qtox?

### ISSUE THREAD

**Issue by phd21** [qTox/qTox#5751]
Hi qTox developers,

Thank you so much for this wonderful multimedia messaging software client.  I recommend this frequently in the Linux Mint forums and elsewhere. I hope you are all doing well.

Question Re: Are there any ways to make file transfers faster through qtox? 

I can easily transfer screenshots and small tiny files, but when trying to send even a small 15mb file, it barely gets to around 20kbps even though my Internet connection and the people I connect with have fairly fast Internet service around 20mbps+ or faster?

Are there any tweaks that can be made to make qTox file transfers faster, if so, what are they? Are qTox file transfers also going through the Tox network and is the Tox network the "bottleneck" (cause of slowness)?

Perhaps changing or adding some faster file transfer options? Can a secure connection be made directly with the qTox user or users currently connected with using P2P, sFTP, SSH, something else, etc...?

I am using Linux Mint 19.x and KDE Neon both are based on Ubuntu 18.04 (64-bit) on a Dell OptiPlex 780 usff computer with a pretty fast Internet connection. 

You are using qTox version v1.16.3-388-gc0fdc42e
Commit hash: c0fdc42e3304a5b6b524ec4469e2b154524ca173
toxcore version: 0.2.8 Qt version: 5.12.0

Best regards,
Phil (phd21)

**Comment by sudden6** [comment]
Unfortunately this is a bug in c-toxcore as far as we know, the problem is nobody seems to know or want to spend time to fix it.

**Comment by anthonybilinski** [comment]
I have a PR that I worked on a bit but I need to go back to for toxcore: https://github.com/TokTok/c-toxcore/pull/1323
It looks like both the RTT calculation is bad, and the congestion control algorithm is bad, but I need to look into it some more and provide a fix. As far as I could tell it's entirely outside of qTox's control - but it's definitely an issue and there's no reason Tox should be limited to such bad speeds.

> Perhaps changing or adding some faster file transfer options? Can a secure connection be made directly with the qTox user or users currently connected with using P2P, sFTP, SSH, something else, etc...?

Tox connections are peer to peer by default, so unless you're using a proxy or have UDP disabled (settings -> advanced -> uncheck "Enable UDP"), there's no bottleneck in the network except for your own network speeds. There's no good reason Toxcore shouldn't be able to saturate any network you give it.

**Comment by phd21** [comment]
Hi sudden6, anthonybilinski,

Thank you both for your prompt replies. I appreciate that.

What does "PR" stand for?

I figured Tox connections are peer to peer, but then why the horrible transfer rates except that there is some bug somewhere in the software regarding file transfers. Okay, if that is the case that there is some "bug" in the software, then please fix it. Thank you again. I would be happy to test any changes anyone comes up with regarding this or anything else with qTox.

Best regards to everyone involved in QTox and Tox in general.
Phil (phd21)

**Comment by anthonybilinski** [comment]
> What does "PR" stand for?

"Pull request", basically just suggested changes. I'll keep this ticket updated when the problem is fixed in toxcore :)

**Comment by phd21** [comment]
Thanks Anthony...

So the file transfer problem(s) are within the Toxcore per se not qTox?

Have a great day and night...

**Comment by sudden6** [comment]
> So the file transfer problem(s) are within the Toxcore per se not qTox?

Yes

**Comment by 2635599** [comment]
i'm with @phd21 on this, does really need to be looked at to see what's causing the bottleneck on transfers.

**Comment by rekkersNZ** [comment]
So i rock a 1000/600 connection on fibre - sending files to another pc inside the network results in 5.36mb, the same speeds i get sending to mates online aswell... This ia total deal-breaker when this app touts p2p.. totally needs to be looked at.. cant expect the app to take off with these kind of problems.. not that im dissing any of the people who put there time and effort into the project, will be watching eagerly..

**Comment by anthonybilinski** [comment]
Agreed, there's no reason a p2p connection shouldn't be able to saturate whatever link it's on, especially given that it's a custom UDP based protocol, it should be able to outperform TCP for file transfers with some packet loss.

I haven't gotten back to looking at the toxcore suspected issue, but I still want to get back to that. It annoys me roughly daily.

**Comment by rekkersNZ** [comment]
Has there been a fix applied for these speed issues as of yet.? I cant believe the core of it hasnt been fixed yet all these other features are added... its like building a car with a cracked chassis then putting all sorts of mods on it.. it will eventually break everything....

**Comment by anthonybilinski** [comment]
No work has been done on it since last update, it still looks like a toxcore issue with its congestion algorithm. PR here https://github.com/TokTok/c-toxcore/pull/1323 still needs to be expanded and finished. I'm currently focussing on other things for qTox, so don't see this happening in the next couple months at least.
